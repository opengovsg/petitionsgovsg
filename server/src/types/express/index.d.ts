declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
