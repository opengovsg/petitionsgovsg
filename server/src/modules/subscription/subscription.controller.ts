import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { CreateSubscriptionReqDto, ErrorDto } from '~shared/types/api'
import { createLogger } from '@/bootstrap/logging'
import { ControllerHandler } from '@/types/response-handler'
import { SubscriptionService } from './subscription.service'

const logger = createLogger(module)

export class SubscriptionController {
  private subscriptionService: Public<SubscriptionService>

  constructor({
    subscriptionService,
  }: {
    subscriptionService: Public<SubscriptionService>
  }) {
    this.subscriptionService = subscriptionService
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @body email
   * @returns 200 with new subscription id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 403 if user is not authorized to answer question
   * @returns 500 if database error
   */
  createSubscription: ControllerHandler<
    { id: string },
    number | ErrorDto,
    CreateSubscriptionReqDto,
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: errors.array()[0].msg })
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      // Save Subscription in the database
      const subscriptionId = await this.subscriptionService.createSubscription({
        email: req.body.email,
        postId: Number(req.params.id),
      })
      return res.status(StatusCodes.OK).json(subscriptionId)
    } catch (error) {
      logger.error({
        message: 'Error while adding new subscription',
        meta: {
          function: 'createSubscription',
          userId: req.user.id,
          postId: req.params.id,
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }
  }
}
