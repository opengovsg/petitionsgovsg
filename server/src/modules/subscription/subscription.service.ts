import type { Sequelize as SequelizeType } from 'sequelize'
import { Subscription } from '~shared/types/base'
import { ModelDef } from '@/types/sequelize'

export class SubscriptionService {
  private Subscription: ModelDef<Subscription>
  private sequelize: SequelizeType

  constructor({
    Subscription,
    sequelize,
  }: {
    Subscription: ModelDef<Subscription>
    sequelize: SequelizeType
  }) {
    this.Subscription = Subscription
    this.sequelize = sequelize
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @param email subscriber's email
   * @returns id of new subscription if it is successfully created
   */
  createSubscription = async ({
    postId,
    email,
  }: Pick<Subscription, 'email' | 'postId'>): Promise<number> => {
    const subscriptionId = await this.sequelize.transaction(
      async (transaction) => {
        const subscription = await this.Subscription.create(
          {
            postId: postId,
            email: email,
          },
          { transaction },
        )
        return subscription.id
      },
    )
    return subscriptionId
  }
}
