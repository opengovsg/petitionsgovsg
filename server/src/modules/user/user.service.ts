import { User } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'
import { LoadPublicUserDto } from '~shared/types/api'
export class UserService {
  private User: ModelDef<User>

  constructor({ User }: { User: ModelDef<User> }) {
    this.User = User
  }

  loadUser = async (userId: number): Promise<LoadPublicUserDto> => {
    return this.User.findByPk(userId) as Promise<LoadPublicUserDto>
  }

  loadFullUser = async (userId: number): Promise<User> => {
    return this.User.findByPk(userId) as Promise<User>
  }

  loadUserBySgid = async (sgid: string): Promise<LoadPublicUserDto> => {
    return this.User.findOne({ where: { sgid: sgid } })
  }
}
