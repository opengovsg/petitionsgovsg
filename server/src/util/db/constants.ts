export const POST_ID = 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847'

export const INVALID_POST_ID = 'af7b6aa9-7c6e-4ea5-b536-4e55bc506840'
