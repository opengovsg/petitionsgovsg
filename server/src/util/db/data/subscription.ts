export const mockSubscription = {
  id: 1,
  email: 'test@email.com',
  postId: 1,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}
