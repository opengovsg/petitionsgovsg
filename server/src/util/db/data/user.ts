import { UserAuthType } from '~shared/types/api'

export const mockUser = {
  id: 'sgid',
  fullname: 'John Doe',
  type: UserAuthType.Public,
}
