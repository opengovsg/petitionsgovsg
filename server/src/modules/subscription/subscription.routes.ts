import { AuthMiddleware } from '../auth/auth.middleware'
import express from 'express'
import { body } from 'express-validator'
import { SubscriptionController } from './subscription.controller'
import { limiter } from '@/middleware/limiter'

export const routeSubscriptions = ({
  controller,
  authMiddleware,
}: {
  controller: SubscriptionController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()
  const { authenticate } = authMiddleware

  /**
   * Create a subscription attached to a post
   * @route   POST /api/subscriptions/:id
   * @returns 200 with new subscriptions id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.post(
    '/:id',
    [
      authenticate,
      limiter,
      body('email', 'email should be a valid format').isString(),
    ],
    controller.createSubscription,
  )
  return router
}
