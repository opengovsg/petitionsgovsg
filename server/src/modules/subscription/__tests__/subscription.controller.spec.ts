import supertest from 'supertest'
import express from 'express'
import { StatusCodes } from 'http-status-codes'
import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '@/util/db/jest-db'
import { SubscriptionController } from '../subscription.controller'
import { POST_ID } from '@/util/db/constants'
import { ControllerHandler } from '@/types/response-handler'
import { mockUser, mockUserJWT } from '@/util/db/data/user'
import { errAsync } from 'neverthrow'
import { DatabaseError } from '@/modules/core/core.errors'
import { mockSubscription } from '@/util/db/data/subscription'

describe('SubscriptionController', () => {
  let db: SequelizeWithModels

  // Set up auth middleware to inject user
  let authUser: Express.User | undefined
  let jwt: { jwt: string } | undefined
  const middleware: ControllerHandler = (req, _res, next) => {
    req.user = authUser
    req['cookies'] = jwt
    next()
  }

  // Set up express-validator middleware
  const noErrors: { errors: () => { msg: string }[] }[] = []
  let errors: { errors: () => { msg: string }[] }[] = noErrors
  const invalidateIfHasErrors: ControllerHandler = (req, _res, next) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    req['express-validator#contexts'] = errors
    next()
  }

  const subscriptionService = {
    createSubscription: jest.fn(),
  }

  const subscriptionController = new SubscriptionController({
    subscriptionService,
  })

  const postPath = `/${POST_ID}`

  beforeAll(async () => {
    db = await createTestDatabase()
  })

  beforeEach(async () => {
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('createSubscription', () => {
    const app = express()
    app.use(express.json())
    app.use(middleware)
    app.use(invalidateIfHasErrors)
    app.post('/:id', subscriptionController.createSubscription)
    const request = supertest(app)

    beforeEach(async () => {
      authUser = mockUser
      jwt = mockUserJWT
      errors = noErrors
      subscriptionService.createSubscription.mockResolvedValue(
        mockSubscription.id,
      )
    })

    it('returns BAD_REQUEST on invalidated request', async () => {
      errors = [
        {
          errors: () => [
            {
              msg: 'Validation Error',
            },
          ],
        },
      ]

      const response = await request.post(postPath).send({ email: 'bademail' })

      expect(subscriptionService.createSubscription).not.toHaveBeenCalled()
      expect(response.status).toEqual(StatusCodes.BAD_REQUEST)
      expect(response.body).toMatchSnapshot()
    })

    it('returns UNAUTHORIZED on no user', async () => {
      authUser = undefined
      jwt = undefined

      const response = await request
        .post(postPath)
        .send({ email: mockSubscription.email })

      expect(subscriptionService.createSubscription).not.toHaveBeenCalled()
      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toStrictEqual({ message: 'User not signed in' })
      expect(response.body).toMatchSnapshot()
    })

    it('returns OK on valid submission', async () => {
      const subscriptionAttributes = {
        email: mockSubscription.email,
        postId: mockSubscription.postId,
      }

      const response = await request
        .post(postPath)
        .send({ email: mockSubscription.email })

      expect(subscriptionService.createSubscription).toHaveBeenCalledWith(
        subscriptionAttributes,
      )
      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(mockSubscription.id)
      expect(response.body).toMatchSnapshot()
    })

    it('returns INTERNAL_SERVER_ERROR on bad service', async () => {
      subscriptionService.createSubscription.mockRejectedValue(
        errAsync(new DatabaseError()),
      )
      const subscriptionAttributes = {
        email: mockSubscription.email,
        postId: mockSubscription.postId,
      }

      const response = await request
        .post(postPath)
        .send({ email: mockSubscription.email })

      expect(subscriptionService.createSubscription).toHaveBeenCalledWith(
        subscriptionAttributes,
      )
      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })
      expect(response.body).toMatchSnapshot()
    })
  })
})
