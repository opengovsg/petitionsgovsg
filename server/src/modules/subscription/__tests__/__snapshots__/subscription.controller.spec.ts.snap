// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionController createSubscription returns BAD_REQUEST on invalidated request 1`] = `Object {}`;

exports[`SubscriptionController createSubscription returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`SubscriptionController createSubscription returns OK on valid submission 1`] = `1`;

exports[`SubscriptionController createSubscription returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;
